File: code/javascript/interview_arena_article-01.js
/**
 * Algorithm is like building Lego: hand-made LRU algorithm
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return LRUCache(...args);
}

var LRUCache = function(capacity) {
    this.cap = capacity;
    this.cache = new Map();
};
LRUCache.prototype.get = function(key) {
    if (!this.cache.has(key)) {
        return -1;
    }
    this.makeRecently(key);
    return this.cache.get(key);
};
LRUCache.prototype.put = function(key, value) {
    if (this.cache.has(key)) {
        this.cache.set(key, value);
        this.makeRecently(key);
        return;
    }
    if (this.cache.size >= this.cap) {
        const oldestKey = this.cache.keys().next().value;
        this.cache.delete(oldestKey);
    }
    this.cache.set(key, value);
};
LRUCache.prototype.makeRecently = function(key) {
    const val = this.cache.get(key);
    this.cache.delete(key);
    this.cache.set(key, val);
};

File: code/java/interview_arena_article-01.java
import java.util.LinkedHashMap;

/**
 * LRU Cache idea in one file:
 * LinkedHashMap remembers order, and we move recently used keys to the end.
 */
public class Interview_arena_article_01 {
    public Object solve(Object... args) {
        if (args == null || args.length != 1) {
            throw new IllegalArgumentException("Expected: (int capacity)");
        }
        return new LRUCache(toInt(args[0]));
    }

    private int toInt(Object value) {
        if (value instanceof Number) {
            return ((Number) value).intValue();
        }
        return Integer.parseInt(String.valueOf(value));
    }

    private static class LRUCache {
        int cap;
        LinkedHashMap<Integer, Integer> cache = new LinkedHashMap<>();

        public LRUCache(int capacity) {
            this.cap = capacity;
        }

        public int get(int key) {
            if (!cache.containsKey(key)) {
                return -1;
            }
            makeRecently(key);
            return cache.get(key);
        }

        public void put(int key, int val) {
            if (cache.containsKey(key)) {
                cache.put(key, val);
                makeRecently(key);
                return;
            }
            if (cache.size() >= this.cap) {
                int oldestKey = cache.keySet().iterator().next();
                cache.remove(oldestKey);
            }
            cache.put(key, val);
        }

        private void makeRecently(int key) {
            int val = cache.get(key);
            cache.remove(key);
            cache.put(key, val);
        }
    }
}
